online-shop/src/app/add-product-form/add-product-form.component.ts

```diff
@@ -0,0 +1,55 @@
+import { Component, OnInit } from '@angular/core';
+import { FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { ActivatedRoute } from '@angular/router';
+import { random } from 'lodash';
+import { IProduct } from '../IProduct';
+import { ProductService } from '../service/productService';
+
+@Component({
+  selector: 'app-add-product-form',
+  templateUrl: './add-product-form.component.html',
+  styleUrls: ['./add-product-form.component.css']
+})
+export class AddProductFormComponent implements OnInit {
+
+  myForm: FormGroup;
+  product : IProduct;
+
+  constructor(private fb: FormBuilder, private productService: ProductService) { }
+
+  ngOnInit(): void {
+
+    this.myForm = this.fb.group({
+      name: ['',{validators: [
+                 Validators.required
+      ]}],
+      price: ['',{validators: [
+                  Validators.required,
+                  Validators.pattern('[0-9]*')]}],
+      category:  ['',{validators: [
+        Validators.required,
+        Validators.pattern('[a-zA-Z]*')]}]
+    })
+  }
+
+  get name(){
+    return this.myForm.get('name');
+  }
+
+  get price(){
+    return this.myForm.get('price');
+  }
+
+  get category(){
+    return this.myForm.get('category');
+  }
+ 
+  addProduct(){
+    this.product = {name: this.myForm.value.name,
+                    category: this.myForm.value.category,
+                    price: this.myForm.value.price,
+                    id : random()};
+    this.productService.addProduct(this.product).subscribe(() => alert("Product added succesfully"));
```

Comment:
**suggestion:** Unsubscribe to prevent memory leaks in your code. https://www.thisdot.co/blog/best-practices-for-managing-rxjs-subscriptions